app/manifest_lock_digests_cmd.go

```diff
@@ -0,0 +1,167 @@
+package app
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/alecthomas/hcl"
+	"github.com/cashapp/hermit/cache"
+	"github.com/cashapp/hermit/manifest"
+
+	"github.com/cashapp/hermit"
+	"github.com/cashapp/hermit/errors"
+	"github.com/cashapp/hermit/state"
+	"github.com/cashapp/hermit/ui"
+)
+
+type lockDigestsCmd struct {
+}
+
+func (e *lockDigestsCmd) Help() string {
+	return `
+	This command will build a list of installed packages and there checksums(SHA256) values.
+	
+	That file can be manually reviewed by the project owner and checked in to the repo.
+
+	On subsequent uses of the hermit environment, the installed dependencies will have a reference checksum
+
+	value for hermit to validate against.
+	`
+}
+func (e *lockDigestsCmd) Run(l *ui.UI, env *hermit.Env, cache *cache.Cache, state *state.State) error {
+	//time.Sleep(time.Second * 7)
+	installed, err := env.ListInstalledReferences()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if len(installed) == 0 {
+		fmt.Printf("There are no packages installed. Nothing to do here.")
+		return nil
+	}
+	for _, ref := range installed {
+		task := l.Task(ref.String())
+		pkg, err := env.Resolve(l, manifest.ExactSelector(ref), false)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		//infer the manifest file
+		res, err := env.GetResolver(l, manifest.ExactSelector(ref), true)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		localmanifest, err := res.ResolveManifest(l, manifest.ExactSelector(ref))
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		// parse through all the version blocks and try to populate the hashes
+		if localmanifest.Manifest.SHA256Sums == nil {
+			localmanifest.Manifest.SHA256Sums = make(map[string]string)
+		}
+		for _, mc := range localmanifest.Manifest.Versions {
+			//multiple platform definitions
+			if len(mc.Layer.Platform) > 0 {
```

Comment:
The approach you've taken here of manually walking through each code path duplicates existing code in Hermit, which will make future maintenance and refactoring more difficult. Instead you should  iterate through all core platforms (available as `platform.Core`), resolving the URLs through the existing code that does that. This might be slow and possibly need some optimisations, but this should be relatively infrequent, and will simplify maintenance.

In pseudo-Python-ish-code, something like this:

```
for version in pkg.versions():
  for (os, arch) in platform.Core:
    url = pkg.resolve(os, arch).url # Continue if this combination can't be resolved, as this can happen
```

Reply:
This turned out to be way easier than what I did. Great idea 💡 .
All the duplicate code is gone too now.